lib/urlwatch/jobs.py

```diff
@@ -284,6 +284,11 @@ def retrieve(self, job_state):
         # Save ETag from response into job_state, which will be saved in cache
         job_state.etag = response.headers.get('ETag')
 
+        # If 'pdf2text' filter is selected, return content in bytes instead of in unicode
+        # as that's what's required by the library used by that filter
+        if self.filter and ('pdf2text' in self.filter or any('pdf2text' in subfilter for subfilter in self.filter)):
+            return response.content
+
```

Comment:
"subfilter" is what we call the "parameters" of the filter. Here's my proposal:

- Add `__uses_bytes__ = True` to `Pdf2TextFilter`

Then, to get all filter names:

    bytes_filters = [name for name, class_ in FilterBase.__subclasses__.items()
                     if getattr(class_, '__uses_bytes__', False)]

(by default this should evaluate to `['pdf2text']` if `__uses__bytes__ = True` is set on the `Pdf2TextFilter` class)

Then you can do:

    for filter_name in bytes_filters:
        if self.filter and (filter_name in self.filter or any(filter_name in filter for filter in self.filter)):
            # a bytes filter is selected, return content in bytes instead of in unicode
            # as that's what's required by the library used by that filter
            return response.content

Of course, it's tricky if you try to chain filters... This assumes that the `pdf2text` filter is the "first" filter. The following filter chain won't work (properly):

```
    filter:
        - html2text
        - pdf2text
```

So maybe even check if it's the only filter or if `isinstance(self.filter, list)` then only check if the first filter is a bytes filter (and maybe also check that all others filters are NOT bytes filters).

Reply:
Checking the first filter makes more sense since Job only creates data to the first filter

The issue with chaining only makes sense if you have filters that output bytes, which you don't have.  Also, I can't think what would be a use case for one of them.  Will document that pdf2text needs to be the *first* filter and work on throwing an error if it's fed a str type, which would happen if it's not the first filter.